servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/AmbSingles.java

```diff
@@ -90,13 +91,22 @@ public void onSubscribe(final Cancellable cancellable) {
 
         @Override
         public void onSuccess(@Nullable final T result) {
+            propagateCancel = false;
             state.trySuccess(result);
         }
 
         @Override
         public void onError(final Throwable t) {
+            propagateCancel = false;
             state.tryError(t);
         }
+
+        @Override
+        public void cancel() {
+            if (propagateCancel) {
```

Comment:
@idelpivnitskiy did I understand you correctly that it needs to be both ways, not just the direction it is done here?

Reply:
should we make the state around `propagateCancel` atomic via AtomicReferenceUpdater? otherwise will this be racy (unless we assume synchronization/sequencing is done externally)?

Reply:
Cancellation after terminal Subscriber method is OK.

https://github.com/reactive-streams/reactive-streams-jvm?tab=readme-ov-file#1.6
> If a Publisher signals either onError or onComplete on a Subscriber, that Subscriber’s Subscription MUST be considered cancelled.

https://github.com/reactive-streams/reactive-streams-jvm?tab=readme-ov-file#3.5
> Subscription.cancel MUST respect the responsivity of its caller by returning in a timely manner, MUST be idempotent and MUST be [thread-safe](https://github.com/apple/servicetalk/pull/3040#term_thread-safe).

Reply:
@Scottmitch the way I read the posted rules doesn't mean that it's ok to explicitly call `cancel()` after terminal, it only says that subscription must be **considered** cancelled. 

The rule 2.3 clarifies it clearly, that we should not call `cancel()` after we got terminal.

@daschl yes, protection as needed both ways, the say Scott described in his first message. I will look around if we can reuse any pre-existing atomic.

Reply:
the "idempotent" from 3.5 allows for multiple calls to cancel. 3.7 also talks to this but is superseded by 3.5.

2.3 says that after the Subscriber is terminated via onComplete/onError, that it shouldn't call Subscription.cancel(). However calling `cancel()` doesn't need to strictly prevent Subscriber onComplete/onError.

https://github.com/reactive-streams/reactive-streams-jvm?tab=readme-ov-file#2.3

> Subscriber.onComplete() and Subscriber.onError(Throwable t) MUST NOT call any methods on the Subscription or the Publisher.

https://github.com/reactive-streams/reactive-streams-jvm?tab=readme-ov-file#3.7

> After the Subscription is cancelled, additional Subscription.cancel() MUST be [NOPs](https://github.com/apple/servicetalk/pull/3040#term_nop).

https://github.com/reactive-streams/reactive-streams-jvm?tab=readme-ov-file#2.8

> A Subscriber MUST be prepared to receive one or more onNext signals after having called Subscription.cancel() if there are still requested elements pending [see [3.12](https://github.com/apple/servicetalk/pull/3040#3.12)]. Subscription.cancel() does not guarantee to perform the underlying cleaning operations immediately.

https://github.com/reactive-streams/reactive-streams-jvm?tab=readme-ov-file#3.12

> While the Subscription is not cancelled, Subscription.cancel() MUST request the Publisher to eventually stop signaling its Subscriber. The operation is NOT REQUIRED to affect the Subscription immediately.

Reply:
> However calling cancel() doesn't need to strictly prevent Subscriber onComplete/onError (see 2.8 and 3.12).

💯, thanks for clarifying and bringing these points.
So, we should only prevent cancel after terminal but not terminal after cancel.

Reply:
@daschl worth adding a test if don't have one already to make sure we can still receive complete/error (from first/second) after the result on amb operation is cancelled.

Reply:
Thanks for the clarifications! I updated the PR to reuse the volatile from `DelayedCancellable` and added a test to make sure we still see a terminal signal after the cancellation.